[BOM-782] feat: 챌린지 한 줄 코멘트 목록 조회 API 구현

rladmstn opened the pull request:
## 📌 What
<!-- 해결하고자 한 문제를 간결하고 명확하게 서술해주세요. -->
<!-- ex) 로그인 실패 시 에러 메시지가 출력되지 않던 문제 -->
챌린지의 한 줄 코멘트 목록 조회를 구현합니다.

## ❓ Why
<!-- 이 문제를 왜 해결해야 했는지, 어떤 맥락에서 발생했는지 작성해주세요. -->
<!-- ex) 사용자 피드백으로 오류 발생 시 원인 파악이 어렵다는 의견이 있었음 -->
각 팀 별 챌린지 참여 동기부여를 위함힙니다.

## 🔧 How
<!-- 어떤 방식으로 해결했는지 기술적 접근 방법을 설명해주세요. -->
<!-- 주요 변경 사항, 선택한 로직이나 방법에 대한 이유가 있으면 함께 작성해주세요. -->
### API 설명
- 챌린지 ID와 회원 정보로 해당 사용자가 포함된 팀의 코멘트 목록을 조회합니다.
- 기본적으로 최신순과 타이브레이크 시 id 오름차순으로 정렬합니다.
- start와 end 사이에 작성된 코멘트 목록을 조회합니다.
 - 현재 기획상은 날짜 하루에 대한 코멘트 목록을 조회할 것 같은데, 확장성을 고려해 특정 기간을 설정해서 조회할 수 있도록 구성했습니다.

### `ChallengeComment`의 `quotation` 필드 추가
- 하이라이트/메모에 대한 인용구를 사용할 수 있으면 좋겠다는 피드백을 반영해 `ChallengeComment`에 `quotation` 필드가 추가되었습니다.
- 그래서 조회할 때도 필드에 `quotation` 필드가 추가됩니다. (nullable)
- 자세한 코멘트 작성 방식은 `코멘트 작성 API` PR에서 설명드리겠습니다!

### `ConstraintViolationException` 핸들링 추가
- (모루 PR에도 있는 내용과 같습니다.)
- `@Positive`에 대한 예외 검증이 400이 아닌 500으로 발생해 challengeId에 대한 컨트롤러 검증을 위해 handler를 추가했습니다.

## 👀 Review Point (Optional)
<!-- 리뷰어에게 중점적으로 봐주었으면 하는 부분을 작성해주세요. -->
<!-- ex) 예외 처리 방식이 적절한지 확인 부탁드립니다. -->

rladmstn committed:
feat: 챌린지 한 줄 코멘트 목록 조회 API 구현

rladmstn committed:
chore: 챌린지 한 줄 코멘트 테이블에 quotation 필드 추가

rladmstn committed:
test: 챌린지 한 줄 코멘트 목록 조회 테스트 작성

Ryan-Dia commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/controller/ChallengeCommentController.java:
```diff
@@ -0,0 +1,46 @@
+            @LoginMember Member member,
+            @PathVariable @Positive(message = "id는 1 이상의 값이어야 합니다.") Long challengeId,
+            @RequestParam LocalDate start,
+            @RequestParam LocalDate end,
```
p2 

저희 컨벤션상  `@RequestParam`은 2개 이상 부터라 dto로 묶어야 될 것 같아요  

<img width="453" height="280" alt="Image" src="https://github.com/user-attachments/assets/27eb7782-7b86-41bd-8d26-8d8f7474f42e" />

더불어 `@NotNull` 검증하면 더 좋을 것 같습니다.

rladmstn replied:
헉 3개부터라고 기억하고 있었는데 2개였군요..?
잘못 기억하고 있었네요..! 수정하겠습니다 캐치 감사합니다!

[feat: 챌린지 코멘트 목록 조회 시 날짜 param](https://github.com/woowacourse-teams/2025-bom-bom/pull/594/commits/d61dcf1586197224223ff33855abffeb3e57a263) @ModelAttribute[로 통합](https://github.com/woowacourse-teams/2025-bom-bom/pull/594/commits/d61dcf1586197224223ff33855abffeb3e57a263)

Ryan-Dia replied:
`@RequestParam`으로 받는 것은 스프링 mvc가 기본적으로 `@ModelAttribute`로 간주하기 때문에 생략이 가능한 것으로 알고있습니다.  
저희가 아마 생략을 했던 것으로 기억하는데 확인해보시고 결정하시면 될 것 같습니다.

rladmstn replied:
다른 코드들 확인해봤더니 @ModelAttribtue를 명시해두어서 통일되도록 표기했습니다!

Ryan-Dia commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/repository/ChallengeCommentRepository.java:
```diff
@@ -0,0 +1,37 @@
+        FROM ChallengeComment cc
+        JOIN ChallengeParticipant cp ON cc.participantId = cp.id
+        JOIN Member m ON cp.memberId = m.id
+        JOIN Article a ON cc.articleId = a.id
```
p3

member와 article은 삭제(탈퇴/삭제)될 수 있는 데이터인데, 현재처럼 INNER JOIN으로 묶으면 둘 중 하나라도 없을 때 해당 comment가 조회에서 통째로 누락됩니다.
즉 “항상 존재해야 하는 필수 데이터”가 아니기 때문에, 이 경우에는 LEFT JOIN으로 조회 안정성을 확보하는 편이 좋아 보입니다.

댓글은 회원이 탈퇴하거나 아티클이 삭제되어도 기록으로서 유효해야 할 가능성이 높아요.
예를 들어 공유된 글의 댓글/좋아요가 많이 쌓여 상단에 노출되는 상황에서, 작성자가 탈퇴하거나 아티클이 삭제됐다는 이유만으로 댓글이 갑자기 사라지면 사용자 입장에서는 맥락이 끊기고 어색하게 느껴질 수 있습니다.

그래서 member/article는 LEFT JOIN으로 두고, 화면에서는 닉네임/제목이 없는 경우 “탈퇴한 사용자”, “삭제된 글” 같은 대체 표시값을 두는 방향도 함께 고려해보면 좋겠습니다.

rladmstn replied:
생각해보지 못한 케이스였는데 제안 감사합니다!
디코에서 의견나눴다시피, `ChallengeComment`의 `articleId`에 대한 반정규화로 `newsletterId`, `articleTitle`을 갖도록 수정했습니다.
또한, 회원은 탈퇴 가능성으로 인해 일단 LEFT JOIN 결과로 탈퇴한 회원은 닉네임을 null로 넘기고 client측에서 표시 방법을 결정할 수 있도록 구성했습니다.

테이블과 설계가 변경되어서, 변경사항 위주보단 전체 코드에서 `ChallengeComment` 엔티티와 `ChallengeCommentReposiory` 쿼리 위주로 리뷰 한 번 해주시면 좋을 것 같습니다!

[feat: ChallengeComment에 뉴스레터 id, 아티클 제목 저장하도록 수정](https://github.com/woowacourse-teams/2025-bom-bom/pull/594/commits/7f0e161dddb58fe269b18b6a38fb8b1eb44c792d)

[fix: 탈퇴한 회원에 대해 닉네임 null로 반환](https://github.com/woowacourse-teams/2025-bom-bom/pull/594/commits/0ad3f3b711a28335e212ec684957b32aa10093b4)

Choidongjun0830 commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/domain/ChallengeComment.java:
```diff
@@ -35,11 +37,13 @@ public ChallengeComment(
             Long id,
             @NonNull Long articleId,
             @NonNull Long participantId,
+            String quotation,
```
영어 잘하신다

rladmstn committed:
Revert "[BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)"

This reverts commit 0e0ccdd7be5da65a5d9baa6be751ef3bdcd0a4c2.

rladmstn committed:
Revert "Revert "[BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)""

This reverts commit 3fe1fd134904b5970586348f38a1b9961bc5afc7.

Choidongjun0830 commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/repository/ChallengeCommentRepository.java:
```diff
@@ -0,0 +1,37 @@
+        WHERE cp.challengeTeamId = :teamId
+        AND FUNCTION('DATE', cc.createdAt) BETWEEN :startDate AND :endDate
+    """)
+    Page<ChallengeCommentResponse> findAllByTeamInDuration(
```
오 이거 JPQL로 하면 Pageable이 알아서 적용이 되네여
맨날 페이징을 QueryDSL이랑 같이 써서 몰랐네여

rladmstn replied:
저도 이번에 처음 알게된 사실입니다 ㅋㅋㅋㅋ
알아서 적용되는게 참 편하네요

rladmstn committed:
feat: 코멘트 목록 조회 시 DTO에 해당 뉴스레터 구독 여부 추가

Choidongjun0830 commented on backend/bom-bom-server/src/test/java/me/bombom/api/v1/challenge/controller/ChallengeCommentControllerTest.java:
```diff
@@ -0,0 +1,153 @@
+                        .param("end", LocalDate.now().plusDays(1).toString())
+                        .with(org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication(
+                                authToken))
+                        .accept(org.springframework.http.MediaType.APPLICATION_JSON))
```
p5

혹시 여기 import 예쁘게 안되나여??

rladmstn replied:
AI를 사용했던 흔적이.. 반영했습니다 감사합니다!

[test: 테스트 관련 클래스 import 추가](https://github.com/woowacourse-teams/2025-bom-bom/pull/594/commits/d8eb6b5652187665c55c210b4ad63c99360049f2)

Choidongjun0830 commented on backend/bom-bom-server/src/test/java/me/bombom/api/v1/challenge/controller/ChallengeCommentControllerTest.java:
```diff
@@ -0,0 +1,153 @@
+                .getResponse()
+                .getContentAsString();
+
+        assertThat(objectMapper.readTree(content).get("code").asText()).isEqualTo("M008");
```
p4

M008이라는게 의미를 파악하기 어려운거 같아요!
응답 코드 400인지 검사하거나 `ErrorDetail.INVALID_REQUEST_PARAMETER_VALIDATION.getCode()` 로 검사하면 더 좋을거 같습니다!

rladmstn replied:
이것도 AI의 흔적이네요..ㅎㅎ 테스트코드 검토를 제대로 못했나봐요ㅜ
위에 `.andExpect(status().isBadRequest())` 부분이 이미 있어서 이걸로 검증 대체했습니다.
감사합니다!

[test: 컨트롤러 테스트 시 응답코드만 검증하도록 수정](https://github.com/woowacourse-teams/2025-bom-bom/pull/594/commits/38e23b5174040971bad252094fbb522816ccfe18)

Choidongjun0830 commented on backend/bom-bom-server/src/test/java/me/bombom/api/v1/challenge/service/ChallengeCommentServiceTest.java:
```diff
@@ -0,0 +1,184 @@
+                        .participantId(participant.getId())
+                        .quotation("quote")
+                        .comment("우리 팀 댓글")
+                        .build()
```
p4

이거 따로 아래처럼 인스턴스 필드인가로 빼고 ourTeamComment, otherTeamComment로 하고 아래 테스트에서 getter로 가져다 쓰는거 어떠신가여! 혹시 누가 건드리면 깨질까봐서요!

<img width="322" height="71" alt="Image" src="https://github.com/user-attachments/assets/9e5a888b-fd04-4e67-995a-1cac9e894da2" />

rladmstn replied:
좋습니다! 그게 더 나은 테스트인 것 같아요!
감사합니다!

Choidongjun0830 commented on backend/bom-bom-server/src/test/java/me/bombom/api/v1/challenge/service/ChallengeCommentServiceTest.java:
```diff
@@ -0,0 +1,184 @@
+
+        // then
+        assertThat(result.getTotalElements()).isEqualTo(1);
+        assertThat(result.getContent().get(0).comment()).isEqualTo("우리 팀 댓글");
```
위 코멘트 여기서 getter로 하는거 어떠냐구 한겁니다! ㅎㅎ

Choidongjun0830 reviewed (approved):
새로이 코멘트만 꼼꼼하게 검토해주시면 될거 같아요!
제가 남긴 코멘트는 테스트 개선 위주라서 우선순위는 낮습니다!

Choidongjun0830 committed:
[BOM-780] refactor: 지난 뉴스레터 관리 Role 기반 관리로 수정 (#590)

* refactor: 지난 아티클 아카이빙을 Role ID 기반으로 동작하도록 수정

* test: 지난 아티클 아카이빙을 Role ID 기반 테스트 작성

* feat: RoleRepository 추가

* refactor: 아카이빙 관리자 id 조회 로직 수정

* style: 개행 수정

rladmstn committed:
feat: 챌린지 코멘트 목록 조회 시 날짜 param @ModelAttribute로 통합

rladmstn committed:
test: 테스트 관련 클래스 import 추가

rladmstn committed:
test: 컨트롤러 테스트 시 응답코드만 검증하도록 수정

rladmstn committed:
test: ChallengeComment TestFixture 추가

Choidongjun0830 committed:
[BOM-784] feat: 챌린지 상세 조회 API 구현 (#593)

* feat: 챌린지 상세 조회 API 구현

* feat: PathVariable 검증 예외 처리 추가

* feat: 챌린지 상세 조회 컨트롤러 단위 테스트 작성

* feat: 챌린지 상세 조회 서비스 통합 테스트 작성

* fix: API 엔드포인트 수정

* feat: 응답 DTO에 제약 추가

* style: 응답 DTO 네이밍 수정

* refactor: 에러 핸들링 메서드 분리 및 에러 메시지 수정

* style: 테스트 클래스명 수정

* test: API 엔드포인트 수정에 따른 테스트 수정

* chore: 스웨거 수정

* fix: 최소 수료 조건 로직 수정

rladmstn committed:
feat: ChallengeComment에 뉴스레터 id, 아티클 제목 저장하도록 수정

rladmstn committed:
fix: 요청 파라미터 검증 메서드 중복 제거

rladmstn committed:
fix: 탈퇴한 회원에 대해 닉네임 null로 반환

rladmstn committed:
fix: 챌린지 코멘트 응답에 notnull 제약 제거

kysub99 committed:
[BE][BOM-786] feat: 챌린지 전체 조회 API (#592)

* feat: 챌린지 도메인 및 전체 조회 API 개발

* test: 챌린지 도메인 테스트 작성

* chore: 미사용 import 제거

* [BOM-783] feat: 챌린지  관련 도메인 생성 (#591)

* feat: 챌린지 관련 도메인 생성

* fix: unique 제약 수정

* [BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)

* feat: 아티클 자동삭제 스케줄러 추가

* feat: 아티클 삭제 시 북마크 제외 및 관리자 제한 분기

* test: 아티클 삭제 스케줄러 테스트 작성

* chore: 버저닝 오타 수정

* fix: 아티클 자동 삭제 시 벌크 연산으로 변경

* style: 불필요한 세미콜론 제거

* fix: 아티클 정리 시 아티클 남겨둘 최소 개수 추가

* feat: 아티클 삭제 시 최소 개수 검증 추가

---------

Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* refactor: ChallengeGrade enum 값명 변경

* style: 코드 컨벤션

* refactor: LocalDate.now() 호출을 필요 시점에 직접 호출하도록 변경

* refactor: 불필요한 else 제거

* refactor: toList 로 변경

* refactor: ChallengeParticipant 파라미터 타입 변경

* refactor: ChallengeNewsletter 조회 로직을 Repository로 이동

* chore: 충돌 해결

* feat: 챌린지 조회시 기수 포함

* feat: 챌린지 뉴스레터 목록 정렬

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: Kim Eunsu <88280787+rladmstn@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

Ryan-Dia reviewed (approved):
수정한 부분 다 확인했습니다.
수고하셨습니다.

kysub99 committed:
[BOM-803] feat: 챌린지 신청 가능 여부 조회 api (#597)

* feat: 챌린지 도메인 및 전체 조회 API 개발

* test: 챌린지 도메인 테스트 작성

* chore: 미사용 import 제거

* [BOM-783] feat: 챌린지  관련 도메인 생성 (#591)

* feat: 챌린지 관련 도메인 생성

* fix: unique 제약 수정

* [BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)

* feat: 아티클 자동삭제 스케줄러 추가

* feat: 아티클 삭제 시 북마크 제외 및 관리자 제한 분기

* test: 아티클 삭제 스케줄러 테스트 작성

* chore: 버저닝 오타 수정

* fix: 아티클 자동 삭제 시 벌크 연산으로 변경

* style: 불필요한 세미콜론 제거

* fix: 아티클 정리 시 아티클 남겨둘 최소 개수 추가

* feat: 아티클 삭제 시 최소 개수 검증 추가

---------

Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* refactor: ChallengeGrade enum 값명 변경

* style: 코드 컨벤션

* refactor: LocalDate.now() 호출을 필요 시점에 직접 호출하도록 변경

* refactor: 불필요한 else 제거

* refactor: toList 로 변경

* refactor: ChallengeParticipant 파라미터 타입 변경

* refactor: ChallengeNewsletter 조회 로직을 Repository로 이동

* chore: 충돌 해결

* feat: 챌린지 조회시 기수 포함

* feat: 챌린지 뉴스레터 목록 정렬

* feat: 챌린지 신청 권한 확인 API

* test: Challenge 서비스 테스트

* test: Challenge 컨트롤러 단위 테스트

* refactor: repository에서 챌린지 신청 여부 확인

* [BE][BOM-786] feat: 챌린지 전체 조회 API (#592)

* feat: 챌린지 도메인 및 전체 조회 API 개발

* test: 챌린지 도메인 테스트 작성

* chore: 미사용 import 제거

* [BOM-783] feat: 챌린지  관련 도메인 생성 (#591)

* feat: 챌린지 관련 도메인 생성

* fix: unique 제약 수정

* [BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)

* feat: 아티클 자동삭제 스케줄러 추가

* feat: 아티클 삭제 시 북마크 제외 및 관리자 제한 분기

* test: 아티클 삭제 스케줄러 테스트 작성

* chore: 버저닝 오타 수정

* fix: 아티클 자동 삭제 시 벌크 연산으로 변경

* style: 불필요한 세미콜론 제거

* fix: 아티클 정리 시 아티클 남겨둘 최소 개수 추가

* feat: 아티클 삭제 시 최소 개수 검증 추가

---------

Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* refactor: ChallengeGrade enum 값명 변경

* style: 코드 컨벤션

* refactor: LocalDate.now() 호출을 필요 시점에 직접 호출하도록 변경

* refactor: 불필요한 else 제거

* refactor: toList 로 변경

* refactor: ChallengeParticipant 파라미터 타입 변경

* refactor: ChallengeNewsletter 조회 로직을 Repository로 이동

* chore: 충돌 해결

* feat: 챌린지 조회시 기수 포함

* feat: 챌린지 뉴스레터 목록 정렬

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: Kim Eunsu <88280787+rladmstn@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* feat: 챌린지 신청 권한 확인 API

* test: Challenge 서비스 테스트

* test: Challenge 컨트롤러 단위 테스트

* refactor: repository에서 챌린지 신청 여부 확인

* refactor: 챌린지 뉴스레터 구독 확인 쿼리 단순화

* style: 컨벤션

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: Kim Eunsu <88280787+rladmstn@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

rladmstn committed:
feat: 챌린지 한 줄 코멘트 목록 조회 API 구현

# Conflicts:
#	backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/repository/ChallengeParticipantRepository.java

rladmstn committed:
test: 챌린지 참여자 TestFixture 추가

rladmstn committed:
fix: conflict 해결

rladmstn committed:
Merge branch 'server-dev' into feat-BOM-782/get-challenge-comments
